### DISCUSSION TITLE
Using and importing helpers in conftest and test files

### DISCUSSION THREAD

**Discussion by g0di** [pytest-dev/pytest#11274] [Q&A]
Quite recently I started to get some errors from VSCode failing at discovering my tests because of something like this:
```
2023-08-02 17:54:03.860 [error] Error discovering pytest tests:
 [n [Error]: ImportError while loading conftest 'c:\Users\TBEBGOD\Developement\...\tests\conftest.py'.
tests\conftest.py:10: in <module>
    from tests.helpers import TestDataBuilder
E   ModuleNotFoundError: No module named 'tests'
```

I'm using a src layout with tests in their own folder looking like this:
```
│   pyproject.toml
│       
├───src
│   │   
│   ├───data_api
│   │   │   ...
│           
└───tests
    │   conftest.py
    │   helpers.py
    │   test_e2e.py
    │   __init__.py
```
My pytest config is the following:
```toml
#pyproject.toml
[tool.pytest.ini_options]
addopts = [
    "--verbose",               
    "--durations=0",    
    "--import-mode=importlib"
]
```

If I remove the `--import-mode=importlib` the discovery is working back again.
If I execute the tests using `python -m pytest` switch, tests are running fine as well
>Whether I'm using importlib as import mode or

From what I understand, when running `pytest` directy with `importlib` import mode, pytest does not add the `tests` folder in the path causing that error.

At the moment my question is more: What is the best/recommended way for importing helpers inside our tests considering:
- Reverting to import mode `append` feels to not be the solution as it is recommended to use `importlib`
- Running my test through `python -m pytest` can not be done through VSCode python extension which use a custom script to discover tests: `2023-08-02 18:11:41.126 [info] > .\.venv\Scripts\python.exe ~\.vscode\extensions\ms-python.python-2023.12.0\pythonFiles\testing_tools\run_adapter.py discover pytest -- --rootdir . -s --cache-clear .` so I loose VSCode support for running my tests
- Extracting my helpers into my production code broke clear separation between tests and tested code
- Modifying the python path through `sys.path.append(...)` to add my helpers seems terrible (no type checking, lot of chances to broke this on file changes ...)
- Changing my imports from `from tests.helpers import ...` to `from .helpers import ...` or `from helpers import ...` does not work either
- Having a `__init__.py` in tests folder is problematic because I can import my test code inside my production code
- Removing the `__init__.py` from `tests` folder and putting my helpers inside a `tests/helpers/__init__.py` does not work either using all kind of impots I could do `from .helpers ...` `from helpers ...` `from tests.helpers ...` 

Thank you for your help ;)

**Comment by g0di** [comment]
One year later, I'm still relying on `prepend` import-mode and `tests` folder being a package (`__init__.py` at root). It does the job but I'm still wondering how people actually manage that, if anyone want to share.

**Comment by nkasing** [comment]
Chanced upon this thread searching for a different issue, and may have a tip for you (2 years later 😆 ).

It's not ideal, but essentially everything needs to be shoved into `conftest.py`. The closest mimic to module behavior is a `Helper` class with a bunch of static/class methods, depending on whether you have interdependencies in your original `helper.py` module.

See [related StackO thread](https://stackoverflow.com/questions/33508060/create-and-import-helper-functions-in-tests-without-creating-packages-in-test-di).

**Reply by g0di** [reply]
Thank you for your comment. This is something I indeed already saw in the past. My main issue with this is about type checking and auto completion by the IDE because you can not import your helper class for typings purposes. 

I ended up doing it the way I describe in my previous comment because it’s the more flexible and easy for me. The only counterpart is that I could import stuff from my tests into my production code but this actually never happened and I feel this as an Ok sacrifice.
